### PR TITLE
Improve the rewriter to simplify more queries

### LINKED ISSUE #14058: Feature Request: Improve Vitess rewriter to simplify queries further.
### Feature Description

For a query like `select 1 from t1 where (false or (false or a in (1)))`, I noticed that Vitess simplifies the query into `select 1 from t1 where false or a in (1)`.

The query, however, can be further simplified into `select 1 from t1 where a in (1)`.

There are benefits of simplifying the query which can further lead to more simplifications and even merging queries together. Vitess in general works better with AND clauses and not so much with ORs, so the more we can eliminate the better.

### Use Case(s)

The simplification of queries is better for better performance, or just in general too even if there is no performance impact.

Also, this is going to be used in a upcoming Foreign key PR :sung

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR improves the rewriter to simplify more queries as pointed out in https://github.com/vitessio/vitess/issues/14058.

After the changes in this PR, when we try to simplify the query `select 1 from t1 where (false or (false or a in (1)))`, we get the query `select 1 from t1 where a in (1)` which is the simplest version possible.

## Related Issue(s)

- Fixes #14058 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/semantics/early_rewriter_test.go [hunk 1]
```diff
@@ -444,3 +444,95 @@ func TestSemTableDependenciesAfterExpandStar(t *testing.T) {
...
+				}, {
+					Name: sqlparser.NewIdentifierCI("b"),
+					Type: sqltypes.VarChar,
+				}, {
+					Name: sqlparser.NewIdentifierCI("c"),
+					Type: sqltypes.VarChar,
+				}},
+				ColumnListAuthoritative: true,
+			},
+			"t2": {
```
Comment (systay) [inline go/vt/vtgate/semantics/early_rewriter_test.go:471]: nitpick: too much copy pasta!

**Review by systay — APPROVED** [review]

**Review by harshit-gangal — APPROVED** [review]

**Comment by harshit-gangal** [comment]
Legit panic
```
2023-09-21T10:20:35.7441050Z --- FAIL: TestFullGeneration (0.78s)
2023-09-21T10:20:35.7441226Z panic: runtime error: invalid memory address or nil pointer dereference [recovered]
2023-09-21T10:20:35.7441394Z 	panic: runtime error: invalid memory address or nil pointer dereference
2023-09-21T10:20:35.7441541Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x6cc33f]
2023-09-21T10:20:35.7441549Z 
2023-09-21T10:20:35.7441629Z goroutine 19 [running]:
2023-09-21T10:20:35.7441739Z testing.tRunner.func1.2({0x74b8e0, 0xa3fc40})
2023-09-21T10:20:35.7441900Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1545 +0x238
2023-09-21T10:20:35.7441987Z testing.tRunner.func1()
2023-09-21T10:20:35.7442141Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1548 +0x397
2023-09-21T10:20:35.7442223Z panic({0x74b8e0?, 0xa3fc40?})
2023-09-21T10:20:35.7442370Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:914 +0x21f
2023-09-21T10:20:35.7442526Z vitess.io/vitess/go/tools/asthelpergen.GenerateASTHelpers(0xc000107ea8)
2023-09-21T10:20:35.7442703Z 	/home/runner/work/vitess/vitess/go/tools/asthelpergen/asthelpergen.go:212 +0x25f
2023-09-21T10:20:35.7442855Z vitess.io/vitess/go/tools/asthelpergen.TestFullGeneration(0xc000085040)
2023-09-21T10:20:35.7443069Z 	/home/runner/work/vitess/vitess/go/tools/asthelpergen/asthelpergen_test.go:28 +0xf1
2023-09-21T10:20:35.7443165Z testing.tRunner(0xc000085040, 0x7d7c98)
2023-09-21T10:20:35.7443319Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1595 +0xff
2023-09-21T10:20:35.7443426Z created by testing.(*T).Run in goroutine 1
2023-09-21T10:20:35.7443613Z 	/opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x3ad
2023-09-21T10:20:35.7443731Z FAIL	vitess.io/vitess/go/tools/asthelpergen	0.794s
```

File: go/vt/vtgate/semantics/early_rewriter_test.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtgate/semantics/early_rewriter_test.go]: Fixed

**Comment by GuptaManan100** [comment]
Fixed the panic

**Pull request merged**